scrapy/crawler.py

```diff
@@ -178,6 +178,51 @@ def stop(self) -> Generator[Deferred, Any, None]:
             assert self.engine
             yield maybeDeferred(self.engine.stop)
 
+    @staticmethod
+    def _get_component(component_class, components):
+        for component in components:
+            if (
+                type(component)  # pylint: disable=unidiomatic-typecheck
+                is component_class
```

Comment:
I think it'd be good to check what's the best behavor here, based on the use cases in the wild: should the default be to match the type exactly, or to use ininstance (i.e. allow subclasses).

Reply:
We could add a boolean parameter to control that, but I guess even then we should check use cases to choose a good default.

Reply:
Use cases I could find in the wild:
- `isinstance`: [1](https://stackoverflow.com/q/64495540/939364), [2](https://stackoverflow.com/a/69681922/939364), [3](https://stackoverflow.com/a/40385860/939364), [4](https://stackoverflow.com/a/33007741/939364), [5](https://stackoverflow.com/a/41994630/939364).
- `type` match: none found.
- `str` in: [1](https://stackoverflow.com/a/54776137/939364), [2](https://github.com/scrapy/scrapy/issues/3590#issuecomment-465796924).
- N/A: [1](https://github.com/scrapy/scrapy/issues/4253#issuecomment-568485213), [2](https://stackoverflow.com/a/60705942/939364), [3](https://stackoverflow.com/a/36623006/939364).

So, do we go with `isinstance`? Do we provide a boolean parameter to switch to a `type` match?

Reply:
I assumed checking the exact type is what people would actually want (not sure if real-world code examples show that they don't, as people could just forget about the subclasses).

Reply:
The use case for isinstance: there is e.g. a middleware (Scrapy built-in, or provided by some package). Some other object (could be e.g. some extension) uses a components getter to find the middleware, and read some information from it. Then, to change the behavior of the middleware, user makes a subclass, disables the original middleware, and enables the new one. 

The question is: should the extension keep reading the information from the new middleware, or should it behave as is there is no middeware available anymore? 

It seems there are arguments for both. For me it looks like using a subclass is better, because subclass is expected to keep the features of the base class, and extend it. So, +0.5 to do isinstance.

Reply:
> It seems there are arguments for both.

Yeah.

> For me it looks like using a subclass is better, because subclass is expected to keep the features of the base class, and extend it.

Makes sense, let's do isinstance then. Shluld we still have a flag to switch it?

Reply:
> Shluld we still have a flag to switch it?

I think we shouldn't do it now, without a real-world use case.